File: pages/LoginPage.ts
import { expect, Locator, Page } from '@playwright/test';
import { BasePage } from './BasePage';
import { env } from '../config/env';

/**
 * Page Object for the CRM authentication (login) page.
 * URL: {baseUrl}/admin/authentication
 */
export class LoginPage extends BasePage {
  readonly path = '/admin/authentication';

  readonly emailInput: Locator;
  readonly passwordInput: Locator;
  readonly rememberMeCheckbox: Locator;
  readonly loginButton: Locator;
  readonly forgotPasswordLink: Locator;
  readonly errorAlert: Locator;

  constructor(page: Page) {
    super(page);
    this.emailInput = page.locator('#email');
    this.passwordInput = page.locator('#password');
    this.rememberMeCheckbox = page.locator('#remember');
    this.loginButton = page.getByRole('button', { name: 'Login' });
    this.forgotPasswordLink = page.getByRole('link', { name: 'Forgot Password?' });
    this.errorAlert = page.locator('#alerts .alert-danger');
  }

  /** Navigate directly to the login page. */
  async open(): Promise<void> {
    await this.goto(`${env.baseUrl}${this.path}`);
    await expect(this.emailInput).toBeVisible();
  }

  async fillEmail(email: string): Promise<void> {
    await this.emailInput.fill(email);
  }

  async fillPassword(password: string): Promise<void> {
    await this.passwordInput.fill(password);
  }

  async checkRememberMe(): Promise<void> {
    await this.rememberMeCheckbox.check();
  }

  async submit(): Promise<void> {
    await this.loginButton.click();
  }

  /** Fill the credentials and submit the form in one step. */
  async login(email: string, password: string, rememberMe = false): Promise<void> {
    await this.fillEmail(email);
    await this.fillPassword(password);
    if (rememberMe) {
      await this.checkRememberMe();
    }
    await this.submit();
  }

  async getErrorMessage(): Promise<string> {
    await expect(this.errorAlert).toBeVisible();
    return (await this.errorAlert.textContent())?.trim() ?? '';
  }
}

File: selenium-java-automation/src/main/java/com/anhtester/crm/pages/LoginPage.java
package com.anhtester.crm.pages;

import com.anhtester.crm.config.ConfigReader;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

/**
 * Page Object for the CRM authentication (login) page.
 * URL: {baseUrl}/admin/authentication
 */
public class LoginPage extends BasePage {

    private static final String PATH = "/admin/authentication";

    private final By emailInput = By.id("email");
    private final By passwordInput = By.id("password");
    private final By rememberMeCheckbox = By.id("remember");
    private final By loginButton = By.cssSelector("button[type='submit']");
    private final By forgotPasswordLink = By.linkText("Forgot Password?");
    private final By errorAlert = By.cssSelector("#alerts .alert-danger");

    public LoginPage(WebDriver driver) {
        super(driver);
    }

    /** Navigate directly to the login page and wait for it to be ready. */
    public LoginPage openPage() {
        open(ConfigReader.baseUrl() + PATH);
        waitForVisible(emailInput);
        return this;
    }

    public LoginPage enterEmail(String email) {
        type(emailInput, email);
        return this;
    }

    public LoginPage enterPassword(String password) {
        type(passwordInput, password);
        return this;
    }

    public LoginPage checkRememberMe() {
        click(rememberMeCheckbox);
        return this;
    }

    public void clickLogin() {
        click(loginButton);
    }

    /** Fills the credentials and submits the form in one step. */
    public void login(String email, String password) {
        enterEmail(email);
        enterPassword(password);
        clickLogin();
    }

    public void clickForgotPassword() {
        click(forgotPasswordLink);
    }

    public boolean isErrorMessageDisplayed() {
        return isVisible(errorAlert);
    }

    public String getErrorMessage() {
        return getText(errorAlert);
    }

    public boolean isLoginButtonDisplayed() {
        return isVisible(loginButton);
    }

    public boolean isAtLoginPage() {
        return getCurrentUrl().contains(PATH);
    }
}
